pkg/template/processor_test.go

```diff
@@ -0,0 +1,345 @@
+
+		// then
+		require.NoError(t, err)
+		require.Len(t, objs, 2)
+
+		// project request
+		verifyResource(t, objs[0].Object.GetObjectKind(), project, commit, user)
+
+		// role binding
+		verifyResource(t, objs[1].Object.GetObjectKind(), project, user)
+	})
+
+	t.Run("template with missing required params", func(t *testing.T) {
+		// given
+		s := addToScheme(t)
+
+		values := make(map[string]string)
+
+		cl := test.NewFakeClient(t)
+		p := template.NewProcessor(cl, s, logger(t))
+
+		// when
+		objs, err := p.Process(templateContent(projectRequestObj, roleBindingObj), values)
+
+		// then
+		require.Error(t, err, "fail to process as not providing required param PROJECT_NAME")
+		assert.Nil(t, objs)
+	})
+
+}
+
+func TestProcessAndApply(t *testing.T) {
+	t.Run("project Request - ok", func(t *testing.T) {
+		//given
+		s := addToScheme(t)
+		project, commit, user := templateVars()
+
+		values := map[string]string{
+			"PROJECT_NAME": project,
+			"COMMIT":       commit,
+			"USER_NAME":    user,
+		}
+
+		cl := test.NewFakeClient(t)
+		p := template.NewProcessor(cl, s, logger(t))
+
+		//when
+		err := p.ProcessAndApply(templateContent(projectRequestObj), values)
+
+		//then
+		require.NoError(t, err)
+		verifyProjectRequest(t, cl, project)
+	})
+
+	t.Run("role binding - ok", func(t *testing.T) {
+		//given
+		s := addToScheme(t)
+		project, commit, user := templateVars()
+
+		values := map[string]string{
+			"PROJECT_NAME": project,
+			"COMMIT":       commit,
+			"USER_NAME":    user,
+		}
+
+		cl := test.NewFakeClient(t)
+		p := template.NewProcessor(cl, s, logger(t))
+
+		//when
+		err := p.ProcessAndApply(templateContent(roleBindingObj), values)
+
+		//then
+		require.NoError(t, err)
+		verifyRoleBinding(t, cl, project)
+	})
+
+	t.Run("project request role binding - ok", func(t *testing.T) {
+		//given
+		s := addToScheme(t)
+		project, commit, user := templateVars()
+
+		values := map[string]string{
+			"PROJECT_NAME": project,
+			"COMMIT":       commit,
+			"USER_NAME":    user,
+		}
+
+		cl := test.NewFakeClient(t)
+		p := template.NewProcessor(cl, s, logger(t))
+
+		//when
+		err := p.ProcessAndApply(templateContent(projectRequestObj, roleBindingObj), values)
+
+		//then
+		require.NoError(t, err)
+		verifyProjectRequest(t, cl, project)
+		verifyRoleBinding(t, cl, project)
+	})
+}
+
+func addToScheme(t *testing.T) *runtime.Scheme {
+	s := scheme.Scheme
+	err := apis.AddToScheme(s)
+	require.NoError(t, err)
+	utilruntime.Must(apitemplate.Install(s)) // see https://github.com/openshift/oc/blob/master/cmd/oc/oc.go#L77
+	return s
+}
+
+func templateContent(objs ...string) []byte {
+	tmpl := templateMetadata
+	for _, obj := range objs {
+		tmpl += obj
+	}
+
+	return []byte(tmpl + templateParams)
+}
+
+func logger(t *testing.T) logr.Logger {
+	zapLog, err := zap.NewDevelopment()
+	require.NoError(t, err)
+	return zapr.NewLogger(zapLog)
+}
+
+func templateVars() (string, string, string) {
```

Comment:
Minor feedback for better usability, incorporate at your will.
This method returns three values and how this three values are used is sort of pre-define.  For ex, first is project, second is commit and third is user.

This method can be enhance to take three input string.  For every blank_string it should generate name with uuid and if string is not blank then return the same string in output with preserve ordering.  With this the usage of this func will be help a lot in your test.

Reply:
Since the function doesn't take any parameter, then I don't see any reason why we should have it. We can easily generate the values as global variables and use them in the tests. I know that it generates the name with the timestamp, but is it really needed?